library(pacman)
p_load(multtest, matrixStats, plyr)
#################################################
load("C:/Users/nitish.mishra/OneDrive - University of Nebraska Medical Center/eMap/RTCGAToolbox/TCGABiolink/GDC-Harmonized/CHOL/CHOL.RData")
## hm450.manifest.hg38.tsv is the annotation file generated by Hui Shen, NAR, 2016 (Comprehensive chracterization, annotation and innovative use of infinium DNA methylation BeadChip)  
rm(list = ls()[!ls() %in% c("CHOL.Meth", "CHOL.FPKM", "CHOL.FPKM.UQ", "CHOL.HTSeq.Counts", "CHOL.miRNA.Counts", "CHOL.miRNA.RPM", "CHOL.Clinical")])
hm450.anno <-read.table("../../HuiShenEPIC-Annotation/hm450.manifest.hg38.tsv",header=T,sep="\t",na.strings="",row.names=NULL,quote="",comment.char="", stringsAsFactors = FALSE, check.names = FALSE)
rownames(hm450.anno) <- hm450.anno$probeID
hm450.anno.unmasked <- hm450.anno[hm450.anno$MASK.general=='FALSE',]
hm450.anno.epig <-read.table("../../HuiShenEPIC-Annotation/hm450.manifest.EpigeneticallySilence_Sort.tsv",header=T,sep="\t",na.strings="NA",row.names=NULL,quote="",comment.char="", stringsAsFactors = FALSE, check.names = FALSE)
## hm450.manifest.EpigeneticallySilence_Sort.tsv is the file for CpG sites with distance from TSS. I generated this file from GDC harmonized DNA methylation file for genCode22
hm450.anno.epig <- hm450.anno.epig[which(hm450.anno.epig$Position_to_TSS!='.'),]
hm450.anno.epig.tss1500 <- hm450.anno.epig[abs(as.numeric(hm450.anno.epig$Position_to_TSS)) <= 1500,]
hm450.anno.epig.tss1500.unmasked <- hm450.anno.epig.tss1500[hm450.anno.epig.tss1500$ProbeID%in%hm450.anno.unmasked$probeID,]
hm450.tss1500 <- hm450.anno.epig.tss1500.unmasked[,c("ProbeID", "Transcript_ID", "Gene_Symbol","Position_to_TSS", "Chromosome")]
colnames(hm450.tss1500) <- gsub("Chromosome","Chr", colnames(hm450.tss1500))
TCGA.Harmonized <- read.table("../../HuiShenEPIC-Annotation/TCGA_GDC_Harmonided_RNASeq-GFT.txt",header=T,sep="\t",na.strings="NA",row.names=NULL,quote="",comment.char="", stringsAsFactors = FALSE, check.names = FALSE)

######################################################
cutoff.1 <- 0.3 # This is the the cutoff for normalrowMeans. Exclude all probe above this cutoff mean beta value.
cutoff.2 <- 0.2 # This is cutoff for hyo in normal. Next step we will select probe which have 75% or more below this cutoff.
cutoff.3 <- 0.90 # Atleast this percent of normal sample have beta below 0.1
cutoff.4 <- 0.3 # This is the cutoff for hypermethylation in tumor.
cutoff.5 <- 0.05 # Atleast this percent of tumor more than 0.3
cutoff.6 <- 0.3 #This is the cutoff for dicotomization of tumor samples.
cutoff.7 <- 0.3 # This is cuoff p-value for correlation
cutoff.8 <- 0.25 # Cutoff for % of probes involved in epigenetic gene silencing
######## DNA methylation and expression data #########
Meth <- CHOL.Meth; colnames(Meth) <- substr(colnames(Meth), 1, 16)
Meth$ProbeID <- rownames(Meth); Meth <- na.omit(Meth) # #Meth <- Meth[complete.cases(Meth),] # Like above command this will also remove all "NA" masked data
Meth.tss1500 <- merge(hm450.tss1500, Meth, by = "ProbeID") 
Tumor.Meth <- Meth.tss1500[,grep("-01A", colnames(Meth.tss1500))]
Normal.Meth <- Meth.tss1500[,grep("-11A", colnames(Meth.tss1500))]
Tumor.Meth$ProbeID <- Meth.tss1500$ProbeID; Normal.Meth$ProbeID <- Meth.tss1500$ProbeID
Tumor.Meth <- unique(Tumor.Meth); Normal.Meth <- unique(Normal.Meth)
rownames(Tumor.Meth) <- Tumor.Meth$ProbeID; rownames(Normal.Meth) <- Normal.Meth$ProbeID
Tumor.Meth$ProbeID <- NULL; Normal.Meth$ProbeID <- NULL

Exp <- CHOL.FPKM; colnames(Exp) <- substr(colnames(Exp), 1, 16)
Exp <- log2(Exp+1)
sample <- intersect(colnames(Tumor.Meth), colnames(Exp)) ##Select only sample which have both expression and DNA methylation data
Tumor.exp <- Exp[,sample]
Normal.exp <- Exp[,grep("-11A", colnames(Exp))] ## For normal expression we will take 
Tumor.Meth <- Tumor.Meth[,sample] ## For tumor select only thos sample which have both expression and methylation data
Exp <- cbind(Normal.exp, Tumor.exp)

### Remove probes which have mean beta > 0.3 and 5% or more normal are methylated (beta > 0.3). 
#Also remove probes for which more than 75% normal have beta < 0.1
keep <- !rowMeans(Normal.Meth) >cutoff.1
Normal.Meth.keep <- Normal.Meth[keep,] ## Remove all probes which have mean beta above 0.3 in normal
Normal.hypo <- (Normal.Meth.keep <= cutoff.2)+0
num.Hypo <- rowSums(Normal.hypo)
Normal.keep <- Normal.Meth.keep[( num.Hypo >= dim(Normal.Meth.keep)[2]*cutoff.3),]#Not less than 75% normal's have beta < 0.1

Tumor.keep <- Tumor.Meth[rownames(Normal.keep),]
Tumor.hyper <- (Tumor.keep >= cutoff.4)+0 ##convert matrix in binary at beta >= 0.3. If needed we can also use 0.25.
num.Hyper <- rowSums(Tumor.hyper)
Tumor.keep <- Tumor.keep[(num.Hyper >= round(dim(Tumor.keep)[2]*cutoff.5)),]#### probe at least 5% hypermethylated
Hyper <- (Tumor.keep >= cutoff.6)+0

##############################################
hm450.tss1500.hyper <- hm450.tss1500[hm450.tss1500$ProbeID%in%rownames(Hyper),]
#Gene_Symbol <- unique(hm450.tss1500.hyper$Gene_Symbol)
Exp$ENSB_ID <- rownames(Exp)
Exp <- merge(Exp, TCGA.Harmonized, by = "ENSB_ID")
### I have to merge with Gene_Symbol, so add Gene_Symbol in Exp dama matrix
#Exp$Transcript_ID <- rownames(Exp)
#Exp.Hyper <- Exp[Exp$Transcript_ID%in%hm450.tss1500.hyper$Transcript_ID, colnames(Hyper)]

###############################################
d2 <- Tumor.keep
d2$ProbeID <- rownames(d2)
d2 <- merge(hm450.tss1500.hyper, d2, by="ProbeID")
##############################################
d3 <- merge(d2, Exp, by = "Gene_Symbol")
d3 <- d3[!c(d3$Chr.x%in% c('chrX', 'chrY')),] ## Remove CheX/ChrY associated probes
d3 <- d3[!c(d3$Chr.y%in% c('chrX', 'chrY')),]
meth <- colnames(d3[,grep("-01A.x", colnames(d3))])
exp <- colnames(d3[,grep("-01A.y", colnames(d3))])
#identical(substring(colnames(meth), 1, 15), substring(colnames(exp), 1, 15))
meth <- as.matrix(d3[,meth])
exp <- as.matrix(d3[,exp])
colnames(meth) <- gsub(".x", '', colnames(meth))
colnames(exp) <- gsub(".y", '', colnames(exp))
hypo <- apply(meth,1, function(x) which(x < 0.3))## Substitute of Hyper.bin
hyper <- apply(meth,1, function(x) which(x >= 0.3))
length(hyper)## 14157
a1 <- matrix(nrow=length(hyper), ncol=dim(meth)[2])
a2 <- matrix(nrow=length(hyper), ncol=dim(exp)[2])
for(i in 1:length(hyper))
{
  a <- unlist(hypo[i])
  a1[i,a] <- exp[i,a]
  b <- unlist(hyper[i])
  a2[i,b] <- exp[i,b]
}
########################################################################
a3 <- matrix(nrow=length(hyper), ncol=dim(meth)[2])
a4 <- matrix(nrow=length(hyper), ncol=dim(meth)[2])
for(i in 1:length(hypo))
{
  a <- unlist(hypo[i])
  a3[i,a] <- meth[i,a]
  b <- unlist(hyper[i])
  a4[i,b] <- meth[i,b]
}
###### calculate correlation between Hyper methylated probe beta value and corresponding gene's expression
ll <- mapply(function(x,y)cor.test(a2[x,],a4[y,], method = "spearman", alternative = "l"),
             1:nrow(a2),
             1:nrow(a4),
             SIMPLIFY=FALSE) ## Few have only one or two value, Pearson's correlation create problem for them
cor.value <- sapply(ll,'[[','estimate')
p.value <- sapply(ll,'[[','p.value')
adjust <- mt.rawp2adjp(p.value, proc=(c("Bonferroni")))
adjust.order <- adjust$adjp[order(adjust$index),]
##################################
d3$hyperExpMean <- rowMeans(a2, na.rm=TRUE)
d3$hyperExpSD <- rowSds(a2, na.rm=TRUE)
d3$hypoExpSD <- rowSds(a1, na.rm=TRUE)
d3$hypoExpMean <- rowMeans(a1, na.rm=TRUE)
d3$quartile10.hypo <- apply(a1, 1, quantile, probs=0.1, na.rm=TRUE)
d3$lower10SD1.64 <- d3$hypoExpMean+(-1.64*d3$hypoExpSD)
d3$totalHyper <- apply(a2, 1, function(x) length(which(!is.na(x))))
ss <- subset(a2 < d3$hypoExpMean) ## Probes corresponding gene in Hyper have expression below mean of Hypo group
d3$totalHyperWithLowerMeanHypo <- rowCounts(ss, na.rm = TRUE, value = TRUE)
ss <- subset(a2 > d3$hypoExpMean) ## Probes corresponding gene in Hyper have expression below mean of Hypo group
d3$totalHyperWithGreaterMeanHypo <- rowCounts(ss, na.rm = TRUE, value = TRUE)
d3$percentageHyper <- (d3$totalHyperWithLowerMeanHypo/d3$totalHyper)*100
d3$corr <- cor.value
d4 <- d3[(which(d3$hyperExpMean < d3$lower10SD1.64  & d3$corr < 0)),]## Probe-gene pair, genes which have expression greater than lower 10% of Unmethylated
d5  <- d4[which(d4$percentageHyper >= 60),] #Probes which have <80% samples in Hypermethlated group have lower expression than mean of Unmethylated group
d5 <- d5[(which(d5$corr < 0)),]
############################################################################
#qq <- table(d4$geneID)
TotalEpiSilence <- table(d5$Gene_Symbol)
TotalProbes <- table(d3$Gene_Symbol)
EpiTable <- as.data.frame(cbind(TotalEpiSilence, TotalProbes))
EpiTable <- EpiTable[which(EpiTable$TotalEpiSilence >=1),]
EpiTable$ID <- rownames(EpiTable)
EpigeneticSlencedGene <- EpiTable[which(EpiTable$TotalEpiSilence >= EpiTable$TotalProbes/2),]
detail.EpigeneticSlencedGene <- d5[d5$Gene_Symbol%in%rownames(EpigeneticSlencedGene),]
############################################################################
EpigeneticGene <- unique(detail.EpigeneticSlencedGene$Gene_Symbol)
EpigeneticGene.symbol <- sapply(strsplit(as.character(EpigeneticGene), split = "\\|"), '[', 1)
#File with more than half probes are epigenetically silenced for each gene
ss <- subset(a2 < d3$hypoExpMean) ## Probes corresponding gene in Hyper have expression below mean of Hypo group
ss1 <- as.data.frame(ss)
ss2 <- as.data.frame(ifelse(ss1=="TRUE", 1, 0))
ss2$ID <- d3$geneID
ss3 <- ddply(ss2, "ID", numcolwise(sum), na.rm=TRUE)
ss3 <- merge(ss3, EpiTable, by="ID")
ss4 <- ifelse(ss3[,2:37]/ss3$TotalEpiSilence >0.5, "TRUE", "FALSE")
rownames(ss4) <- ss3$ID
colnames(ss4) <- names(d3[39:74])
colnames(ss4) <- gsub("\\.y", "", colnames(ss4))
SupplementaryTable <- ss4[rownames(ss4)%in%EpigeneticGene, ]## This file be like Supplementary table in 
## TCGA stomach cancer, but I am also using spearman correlation http://www.nature.com/nature/journal/v513/n7517/full/nature13480.html
apply(SupplementaryTable, 1, table) ## Count Number of TRUE and FALSE in each line
rm(ss1, ss2, ss3, ss4)
####################################################################################
######### GO and Pathway Enrichment analysis by using TCGABiolinks ###########
system.time(ansEA <- TCGAanalyze_EAcomplete(TFname="Epigenetically silenced genes in Cholangiocarcinoma (CHOL)",EpigeneticGene.symbol))
TCGAvisualize_EAbarplot(tf = rownames(ansEA$ResBP), 
                        GOBPTab = ansEA$ResBP,
                        GOCCTab = ansEA$ResCC,
                        GOMFTab = ansEA$ResMF,
                        PathTab = ansEA$ResPat,
                        nRGTab = EpigeneticGene.symbol,
                        nBar = 10,
                        filename = "Epigenetic_silenced_genes_pathways_and_GO_enrichment.pdf")