library(pacman)
p_load(multtest, matrixStats, plyr)
#################################################
load("C:/Users/Nitish/Desktop/Epigenetic Silencing/CHOL.RData")
## hm450.manifest.hg38.tsv is the annotation file generated by Hui Shen, NAR, 2016 (Comprehensive chracterization, annotation and innovative use of infinium DNA methylation BeadChip)  
rm(list = ls()[!ls() %in% c("CHOL.Meth", "CHOL.FPKM", "CHOL.miRNA.RPM", "CHOL.FPKM.UQ", "CHOL.HTSeq.Counts")])
hm450.anno <-read.table("C:/Users/Nitish/Desktop/Epigenetic Silencing/hm450.manifest.hg38.tsv",header=T,sep="\t",na.strings="",row.names=NULL,quote="",comment.char="", stringsAsFactors = FALSE, check.names = FALSE)
rownames(hm450.anno) <- hm450.anno$probeID
hm450.anno.unmasked <- hm450.anno[hm450.anno$MASK.general=='FALSE',]
rm(hm450.anno)
hm450.anno.epig <-read.table("C:/Users/Nitish/Desktop/Epigenetic Silencing/hm450.manifest.EpigeneticallySilence_Sort.tsv",header=T,sep="\t",na.strings="NA",row.names=NULL,quote="",comment.char="", stringsAsFactors = FALSE, check.names = FALSE)
## hm450.manifest.EpigeneticallySilence_Sort.tsv is the file for CpG sites with distance from TSS. I generated this file from GDC harmonized DNA methylation file for genCode22
hm450.anno.epig <- hm450.anno.epig[which(hm450.anno.epig$Position_to_TSS!='.'),]
hm450.anno.epig.tss1500 <- hm450.anno.epig[abs(as.numeric(hm450.anno.epig$Position_to_TSS)) <= 1500,]
hm450.anno.epig.tss1500.unmasked <- hm450.anno.epig.tss1500[hm450.anno.epig.tss1500$ProbeID%in%hm450.anno.unmasked$probeID,]
hm450.tss1500 <- hm450.anno.epig.tss1500.unmasked[,c("ProbeID", "Transcript_ID", "Gene_Symbol","Position_to_TSS", "Chromosome")]
colnames(hm450.tss1500) <- gsub("Chromosome","Chr", colnames(hm450.tss1500))
#TCGA.Harmonized <- read.table("C:/Users/Nitish/Desktop/Epigenetic Silencing/TCGA_GDC_Harmonided_RNASeq-GFT.txt",header=T,sep="\t",na.strings="NA",row.names=NULL,quote="",comment.char="", stringsAsFactors = FALSE, check.names = FALSE)
TCGA.Harmonized <- read.table("C:/Users/Nitish/Desktop/Epigenetic Silencing/TCGA_GDC_Harmonided_Uniq_GTF.txt",header=F,sep="\t",na.strings="NA",row.names=NULL,quote="",comment.char="", stringsAsFactors = FALSE, check.names = FALSE)
colnames(TCGA.Harmonized) <- c("Chr", "Reference", "Transcript", "Start", "End", "Strand", "Transcript_ID", "Gene_Symbol")
masked <- which(TCGA.Harmonized$Chr%in% c("chrX", "chrY", "chrM"))
TCGA.Har.Masked <- TCGA.Harmonized[-masked,]
TCGA.Har.Masked <- TCGA.Har.Masked[,c("Chr", "Transcript_ID", "Gene_Symbol")]

masked <- which(hm450.tss1500$Chr%in% c("chrX", "chrY", "chrM"))
hm450.tss1500.Masked <- hm450.tss1500[-masked,]
hm450.tss1500.Masked.Uniq <- unique(hm450.tss1500.Masked[,c("ProbeID", "Gene_Symbol", "Chr")]) ## Unique gene and probe within +/- 1500 from TSS
######## DNA methylation and expression data #########
Meth <- CHOL.Meth; colnames(Meth) <- substr(colnames(Meth), 1, 16)
Exp <- CHOL.FPKM.UQ; colnames(Exp) <- substr(colnames(Exp), 1, 16)
TumorID <- intersect(colnames(Meth[grep("-01A", colnames(Meth))]), colnames(Exp[grep("-01A", colnames(Exp))]))
NormalID <- intersect(colnames(Meth[grep("-11A", colnames(Meth))]), colnames(Exp[grep("-11A", colnames(Exp))]))
Meth$ProbeID <- rownames(Meth); Meth <- na.omit(Meth) # #Meth <- Meth[complete.cases(Meth),] # Like above command this will also remove all "NA" masked data
Meth.tss1500 <- merge(hm450.tss1500.Masked.Uniq, Meth, by = "ProbeID") 
Meth.tss1500$N.Mean <- rowMeans(Meth.tss1500[,grep("-11A", colnames(Meth.tss1500))])#Mean Beta Normal
Meth.tss1500$T.Mean <- rowMeans(Meth.tss1500[,grep("-01A", colnames(Meth.tss1500))])#Mean Beta Tumor
masked <- which(Meth.tss1500$N.Mean > 0.3)
Meth.tss1500.Maked <- Meth.tss1500[-masked,]

Meth.tss1500.Maked$T.Great.0.3 <- rowSums(Meth.tss1500.Maked[,TumorID] > 0.3)/length(TumorID)*100
Meth.tss1500.Maked$N.Less.0.1 <- rowSums(Meth.tss1500.Maked[,NormalID] < 0.1)/length(NormalID)*100
masked <- which(Meth.tss1500.Maked$T.Great.0.3 >= 5 & Meth.tss1500.Maked$N.Less.0.1 >= 50)
#masked <- which(Meth.tss1500.Maked$T.Great.0.3 > 5)
Meth.tss1500.Maked <- Meth.tss1500.Maked[masked,]
#Tumor.Meth$ProbeID <- NULL; Normal.Meth$ProbeID <- NULL
Exp <- Exp[,c(NormalID, TumorID)]
Exp <- log2(Exp+1)
Exp$Transcript_ID <- rownames(Exp)
Exp.Masked <- merge(Exp, TCGA.Har.Masked, by = "Transcript_ID")
dat.meth.exp <- merge(Meth.tss1500.Maked, Exp.Masked, by = "Gene_Symbol")
dat.Me <- dat.meth.exp[,grep("A.x", colnames(dat.meth.exp))]
dat.Ex <- dat.meth.exp[,grep("A.y", colnames(dat.meth.exp))]
colnames(dat.Me) <- gsub(".x", "", colnames(dat.Me))
colnames(dat.Ex) <- gsub(".y", "", colnames(dat.Ex))
dat.Me.T <- dat.Me[,grep("-01A", colnames(dat.Me))]
dat.Me.N <- dat.Me[,grep("-11A", colnames(dat.Me))]
dat.Ex.T <- dat.Me[,grep("-01A", colnames(dat.Ex))]
dat.Ex.N <- dat.Me[,grep("-11A", colnames(dat.Ex))]

T.Hyper <- vector("numeric", length = dim(dat.Ex)[1])
T.Hypo <- vector("numeric", length = dim(dat.Ex)[1])
Hyper.Exp.Mean <- vector("numeric", length = dim(dat.Ex)[1])
Hypo.Exp.Mean <- vector("numeric", length = dim(dat.Ex)[1])
Hypo.Exp.SD <- vector("numeric", length = dim(dat.Ex)[1])
Hypo.Exp.10.Per <- vector("numeric", length = dim(dat.Ex)[1])
Exp.t.test <- vector("numeric", length = dim(dat.Ex)[1])

for(i in 1:dim(dat.meth.exp)[1]){
  Hyper <- which(dat.Me.T[i,] >= 0.3)
  Hypo <- which(dat.Me.T[i,] < 0.3)
  T.Hyper[i] <- length(Hyper)
  T.Hypo[i] <- length(Hypo)
  Hyper.Exp.Mean[i] <- mean(as.numeric(dat.Ex.T[i, Hyper]))
  Hypo.Exp.Mean[i] <- mean(as.numeric(dat.Ex.T[i, Hypo]), na.rm = TRUE)
  Hypo.Exp.SD[i] <- sd(dat.Ex.T[i, Hypo], na.rm = TRUE)
  Hypo.Exp.10.Per[i] <- apply(dat.Ex.T[i, Hypo],1, quantile, probs=0.1, na.rm = TRUE)
  #tmp <- t.test(dat.Ex.T[i,Hyper], dat.Ex.T[i, Hypo], alternative = 'l')
  #Exp.t.test[i] <- tmp$p.value

}
dat.meth.exp$T.Hyper <- T.Hyper
dat.meth.exp$T.Hypo <- T.Hypo
dat.meth.exp$Hyper.Exp.Mean <- Hyper.Exp.Mean
dat.meth.exp$Hypo.Exp.Mean <- Hypo.Exp.Mean
dat.meth.exp$Hypo.Exp.SD <- Hypo.Exp.SD
dat.meth.exp$SD_1.28 <- Hypo.Exp.Mean+(-1.28*Hypo.Exp.SD) #1.28 for the 20% & 1.68 for the 10%
#dat.meth.exp$quant_0.1 <- Hypo.Exp.10.Per
#dat.meth.exp$t.test.P <-  Exp.t.test
#tmp <- multtest::mt.rawp2adjp(dat.meth.exp$t.test.P, proc = "BH")
#dat.meth.exp$t.test <- tmp$adjp[order(tmp$index), ]

# Percentile Score = Mean + (Z-score * SD) ## For lower 10% Z-score = -1.28,-1.64 for 5% and -1.03 for 15%
#a1 <- as.data.frame(table(dat.meth.exp[(dat.meth.exp$t.test[,1]) < 0.01 & dat.meth.exp$Hyper.Exp.Mean < dat.meth.exp$quant_0.1,]$Gene_Symbol))
#a1 <- as.data.frame(table(dat.meth.exp[(dat.meth.exp$t.test[,1]) < 0.01 & dat.meth.exp$Hyper.Exp.Mean < dat.meth.exp$SD_1.28,]$Gene_Symbol))
a1 <- as.data.frame(table(dat.meth.exp[dat.meth.exp$Hyper.Exp.Mean < dat.meth.exp$SD_1.28,]$Gene_Symbol))
a2 <- as.data.frame(table(dat.meth.exp$Gene_Symbol))
EpiRegTable <- merge(a1, a2, by = "Var1")
EpiRegTable$Ratio <- EpiRegTable$Freq.x/EpiRegTable$Freq.y
EpiTable <- EpiRegTable[EpiRegTable$Ratio >= 0.25,]

rm(list = ls()[!ls() %in% c("EpiRegTable","EpiTable","dat.meth.exp","CHOL.Meth", "CHOL.FPKM", "CHOL.miRNA.RPM", "CHOL.FPKM.UQ", "CHOL.HTSeq.Counts")])
save.image(file = "C:/Users/Nitish/Desktop/Epigenetic Silencing/EpigeneticSilence.RData")
