setwd("F:/OneDrive - University of Nebraska Medical Center/Epigenetic Silencing/")
library(pacman)
p_load(multtest, matrixStats, plyr)
#################################################
load("CHOL.RData")
rm(list = ls()[!ls() %in% c("CHOL.Meth", "CHOL.FPKM", "CHOL.miRNA.RPM", "CHOL.FPKM.UQ", "CHOL.HTSeq.Counts")])
## hm450.manifest.hg38.tsv is the annotation file generated by Hui Shen, NAR, 2016 (Comprehensive chracterization, annotation and innovative use of infinium DNA methylation BeadChip)  
hm450.anno <-read.table("hm450.manifest.hg38.tsv",header=T,sep="\t",na.strings="",row.names=NULL,quote="",comment.char="", stringsAsFactors = FALSE, check.names = FALSE)
rownames(hm450.anno) <- hm450.anno$probeID
hm450.anno.unmasked <- hm450.anno[hm450.anno$MASK.general=='FALSE',]
rm(hm450.anno)
hm450.anno.epig <-read.table("hm450.manifest.EpigeneticallySilence_Sort.tsv",header=T,sep="\t",na.strings="NA",row.names=NULL,quote="",comment.char="", stringsAsFactors = FALSE, check.names = FALSE)
## hm450.manifest.EpigeneticallySilence_Sort.tsv is the file for CpG sites with distance from TSS. I generated this file from GDC harmonized DNA methylation file for genCode22
hm450.anno.epig <- hm450.anno.epig[which(hm450.anno.epig$Position_to_TSS!='.'),]
hm450.anno.epig.tss1500 <- hm450.anno.epig[abs(as.numeric(hm450.anno.epig$Position_to_TSS)) <= 1500,]
hm450.anno.epig.tss1500.unmasked <- hm450.anno.epig.tss1500[hm450.anno.epig.tss1500$ProbeID%in%hm450.anno.unmasked$probeID,]
hm450.tss1500 <- hm450.anno.epig.tss1500.unmasked[,c("ProbeID", "Transcript_ID", "Gene_Symbol","Position_to_TSS", "Chromosome")]
colnames(hm450.tss1500) <- gsub("Chromosome","Chr", colnames(hm450.tss1500))
#TCGA.Harmonized <- read.table("TCGA_GDC_Harmonided_RNASeq-GFT.txt",header=T,sep="\t",na.strings="NA",row.names=NULL,quote="",comment.char="", stringsAsFactors = FALSE, check.names = FALSE)
TCGA.Harmonized <- read.table("TCGA_GDC_Harmonided_Uniq_GTF.txt",header=F,sep="\t",na.strings="NA",row.names=NULL,quote="",comment.char="", stringsAsFactors = FALSE, check.names = FALSE)
colnames(TCGA.Harmonized) <- c("Chr", "Reference", "Transcript", "Start", "End", "Strand", "Transcript_ID", "Gene_Symbol")
masked <- which(TCGA.Harmonized$Chr%in% c("chrX", "chrY", "chrM"))
TCGA.Har.Masked <- TCGA.Harmonized[-masked,]
TCGA.Har.Masked <- TCGA.Har.Masked[,c("Chr", "Transcript_ID", "Gene_Symbol")]
masked <- which(hm450.tss1500$Chr%in% c("chrX", "chrY", "chrM"))
hm450.tss1500.Masked <- hm450.tss1500[-masked,]
hm450.tss1500.Masked.Uniq <- unique(hm450.tss1500.Masked[,c("ProbeID", "Gene_Symbol", "Chr")]) ## Unique gene and probe within +/- 1500 from TSS
######## DNA methylation and expression data #########
Meth <- CHOL.Meth; colnames(Meth) <- substr(colnames(Meth), 1, 16)
Exp <- CHOL.FPKM; colnames(Exp) <- substr(colnames(Exp), 1, 16)
TumorID <- intersect(colnames(Meth[grep("-01A", colnames(Meth))]), colnames(Exp[grep("-01A", colnames(Exp))]))
NormalID <- intersect(colnames(Meth[grep("-11A", colnames(Meth))]), colnames(Exp[grep("-11A", colnames(Exp))]))
Meth <- Meth[,c(NormalID, TumorID)]; Exp <- Exp[,c(NormalID, TumorID)]
Meth$ProbeID <- rownames(Meth); Meth <- na.omit(Meth) # #Meth <- Meth[complete.cases(Meth),] # Like above command this will also remove all "NA" masked data
Meth.tss1500 <- merge(hm450.tss1500.Masked.Uniq, Meth, by = "ProbeID") 
Meth.tss1500$N.Mean <- rowMeans(Meth.tss1500[,grep("-11A", colnames(Meth.tss1500))])#Mean Beta Normal
Meth.tss1500$T.Mean <- rowMeans(Meth.tss1500[,grep("-01A", colnames(Meth.tss1500))])#Mean Beta Tumor
masked <- which(Meth.tss1500$N.Mean > 0.3)
Meth.tss1500.Maked <- Meth.tss1500[-masked,]

Meth.tss1500.Maked$T.Great.0.3 <- rowSums(Meth.tss1500.Maked[,TumorID] > 0.3)/length(TumorID)*100
#Meth.tss1500.Maked$N.Less.0.1 <- rowSums(Meth.tss1500.Maked[,NormalID] < 0.1)/length(NormalID)*100
#masked <- which(Meth.tss1500.Maked$T.Great.0.3 >= 5 & Meth.tss1500.Maked$N.Less.0.1 >= 80)
masked <- which(Meth.tss1500.Maked$T.Great.0.3 > 5)
Meth.tss1500.Maked <- Meth.tss1500.Maked[masked,]
ProbeCount <- as.data.frame(table(Meth.tss1500.Maked$Gene_Symbol)) ## Number of probes for each genes within 1.5 Kb from TSS  (+/- 1.5Kb)
colnames(ProbeCount) <- c("Gene_Symbol", "ProbeCount")

#Tumor.Meth$ProbeID <- NULL; Normal.Meth$ProbeID <- NULL
Exp <- Exp[,c(NormalID, TumorID)]
Exp <- log2(Exp+1)
Exp$Transcript_ID <- rownames(Exp)
Exp.Masked <- merge(Exp, TCGA.Har.Masked, by = "Transcript_ID")
dat.meth.exp <- merge(Meth.tss1500.Maked, Exp.Masked, by = "Gene_Symbol")
dat.Me <- dat.meth.exp[,grep("A.x", colnames(dat.meth.exp))]
dat.Ex <- dat.meth.exp[,grep("A.y", colnames(dat.meth.exp))]
colnames(dat.Me) <- gsub(".x", "", colnames(dat.Me))
colnames(dat.Ex) <- gsub(".y", "", colnames(dat.Ex))
dat.Me.T <- dat.Me[,grep("-01A", colnames(dat.Me))]
dat.Me.N <- dat.Me[,grep("-11A", colnames(dat.Me))]
dat.Ex.T <- dat.Ex[,grep("-01A", colnames(dat.Ex))]
dat.Ex.N <- dat.Ex[,grep("-11A", colnames(dat.Ex))]

T.Hyper <- vector("numeric", length = dim(dat.Ex)[1])
T.Hypo <- vector("numeric", length = dim(dat.Ex)[1])
Hyper.Exp.Mean <- vector("numeric", length = dim(dat.Ex)[1])
Hypo.Exp.Mean <- vector("numeric", length = dim(dat.Ex)[1])
Hypo.Exp.SD <- vector("numeric", length = dim(dat.Ex)[1])
Hyper.Meth.Mean <- vector("numeric", length = dim(dat.Ex)[1])
Hypo.Meth.Mean <- vector("numeric", length = dim(dat.Ex)[1])
Hypo.Exp.10.Per <- vector("numeric", length = dim(dat.Ex)[1])
Exp.t.test <- vector("numeric", length = dim(dat.Ex)[1])
Exp.rho <- vector("numeric", length = dim(dat.Ex)[1])
Exp.rho.p <- vector("numeric", length = dim(dat.Ex)[1])

for(i in 1:nrow(dat.Me.T)){
  Hyper <- which(dat.Me.T[i,] >= 0.3)
  Hypo <- which(dat.Me.T[i,] < 0.3)
  T.Hyper[i] <- length(Hyper)
  T.Hypo[i] <- length(Hypo)
  Hyper.Exp.Mean[i] <- mean(as.numeric(dat.Ex.T[i, Hyper]))
  Hypo.Exp.Mean[i] <- mean(as.numeric(dat.Ex.T[i, Hypo]), na.rm = TRUE)
  Hypo.Exp.SD[i] <- sd(dat.Ex.T[i, Hypo], na.rm = TRUE)
  Hyper.Meth.Mean <- mean(as.numeric(dat.Me.T[i, Hyper]))
  Hypo.Meth.Mean <- mean(as.numeric(dat.Me.T[i, Hypo]))
  #Hypo.Exp.10.Per[i] <- apply(dat.Ex.T[i, Hypo],1, quantile, probs=0.1, na.rm = TRUE)
  #tmp <- t.test(as.numeric(dat.Ex.T[i,Hyper]), as.numeric(dat.Ex.T[i, Hypo], alternative = 'l'))
  #Exp.t.test[i] <- tmp$p.value
  tmp <- cor.test(as.numeric(dat.Me.T[i,Hyper]), as.numeric(dat.Ex.T[i,Hyper]), method = "spearman")
  Exp.rho[i] <- tmp$estimate
  Exp.rho.p[i] <- tmp$p.value
}
dat.meth.exp$T.Hyper <- T.Hyper
dat.meth.exp$T.Hypo <- T.Hypo
dat.meth.exp$Hyper.Exp.Mean <- Hyper.Exp.Mean
dat.meth.exp$Hypo.Exp.Mean <- Hypo.Exp.Mean
dat.meth.exp$Hypo.Exp.SD <- Hypo.Exp.SD
dat.meth.exp$Hyper.Meth.Mean <- Hyper.Meth.Mean
dat.meth.exp$Hypo.Meth.Mean <- Hypo.Meth.Mean
dat.meth.exp$rho <- Exp.rho
dat.meth.exp$rho.p <- Exp.rho.p
dat.meth.exp$rho.p.adjust <- p.adjust(Exp.rho.p, "BH")
dat.meth.exp$SD_1.64 <- Hypo.Exp.Mean+(-1.64*Hypo.Exp.SD)
dat.meth.exp$Z.score <- (dat.meth.exp$Hyper.Exp.Mean - dat.meth.exp$Hypo.Exp.Mean)/dat.meth.exp$Hypo.Exp.SD
#1.64 for the 5% & 1.28 for the 10%
#dat.meth.exp$quant_0.1 <- Hypo.Exp.10.Per
#dat.meth.exp$t.test.P <-  Exp.t.test
#tmp <- p.adjust(dat.meth.exp$t.test.P, "BH")
#dat.meth.exp$t.test <- tmp$adjp[order(tmp$index), ]
# Percentile Score = Mean + (Z-score * SD) ## For lower 10% Z-score = -1.64,-1.64 for 5% and -1.03 for 15%

### Either I using Z.score <= -1.64 or criteria which I am using below both are same
dat.meth.exp.SD_1.64 <- dat.meth.exp[which(dat.meth.exp$Hyper.Exp.Mean < dat.meth.exp$SD_1.64),]
dat.meth.exp.SD_1.64 <- na.omit(dat.meth.exp.SD_1.64)
dat.Me.SD <- dat.meth.exp.SD_1.64[,grep("A.x", colnames(dat.meth.exp.SD_1.64))]
dat.Ex.SD <- dat.meth.exp.SD_1.64[,grep("A.y", colnames(dat.meth.exp.SD_1.64))]
#adjust <- mt.rawp2adjp(Exp.rho.p, proc=(c("BH")))
#adjust.order <- adjust$adjp[order(adjust$index),]

colnames(dat.Me.SD) <- gsub(".x", "", colnames(dat.Me.SD))
colnames(dat.Ex.SD) <- gsub(".y", "", colnames(dat.Ex.SD))
dat.Me.SD.T <- dat.Me.SD[,grep("-01A", colnames(dat.Me.SD))]
dat.Ex.SD.T <- dat.Ex.SD[,grep("-01A", colnames(dat.Ex.SD))]

matrix1 <- matrix(nrow = nrow(dat.Ex.SD.T), ncol = ncol(dat.Ex.SD.T))
for(i in 1:dim(dat.Ex.SD.T)[1]){
  Hyper <- which(dat.Me.SD.T[i,] >= 0.3)
  Hypo <- which(dat.Me.SD.T[i,] < 0.3)
  matrix1[i,] <- ifelse(dat.Me.SD.T[i,] >= 0.3 & dat.Ex.SD.T[i,] < as.numeric(rowMeans(dat.Ex.SD.T[i,Hypo])), 1, 0)
}
colnames(matrix1) <- colnames(dat.Ex.SD.T) 
matrix1 <- as.data.frame(matrix1)
matrix1$Gene_Symbol <- dat.meth.exp.SD_1.64$Gene_Symbol
matrix1$ProbeID <- dat.meth.exp.SD_1.64$ProbeID
matrix1 <- matrix1[rowSums(matrix1[,1:36]) >=1,]
matrix1.1 <- matrix1[,1:37]
matrix2 <- aggregate(. ~ Gene_Symbol, data = matrix1.1, sum) ## Sum each sample with Gene_Symbol
## If one sample have two probe it will some with Gene_Symbol
matrix2 <- merge(matrix2, ProbeCount, by= "Gene_Symbol") ## How many probe for each genes
tmp <- matrix2[,2:37]
rownames(tmp) <- matrix2$Gene_Symbol
tmp <- tmp/as.numeric(matrix2$ProbeCount)
tmp <- tmp[apply(tmp[,], MARGIN = 1, function(x) any(x >= 0.5)), ]
rownames(tmp[rowSums(tmp > 0.5) >= 0.05*ncol(tmp), ]) ### Gene with atleast 5% samples with epigenetic silencing
Epigenetic.silenced <- ifelse(tmp >= 0.5, TRUE,FALSE)

rm(list = ls()[!ls() %in% c("Epigenetic.silenced", "matrix1", "matrix2","dat.meth.exp","CHOL.Meth", "CHOL.FPKM", "CHOL.FPKM.UQ")])
save.image(file = "EpigeneticSilence.Nov-06-2017.RData")
