## R code for Epigenetic silencing as suggested by Zheua Wang et. al. 
#lncRNA Epigenetic Landscape Analysis Identifies EPIC1 as an Oncogenic lncRNA that Interacts with MYC and Promotes Cell-Cycle Progression in Cancer. Cancer Cell, Volume 33, Issue 4, 9 April 2018, Pages 706-720.e9
setwd("F:/OneDrive - University of Nebraska Medical Center/Epigenetic Silencing/")
library(pacman)
p_load(multtest, matrixStats, plyr)
#######################################################
load("CHOL.RData")
rm(list = ls()[!ls() %in% c("CHOL.Meth", "CHOL.FPKM", "CHOL.miRNA.RPM", "CHOL.FPKM.UQ", "CHOL.HTSeq.Counts")])
## hm450.manifest.hg38.tsv is the annotation file generated by Hui Shen, NAR, 2016 (Comprehensive chracterization, annotation and innovative use of infinium DNA methylation BeadChip)  
hm450.anno <-read.table("hm450.manifest.hg38.tsv",header=T,sep="\t",na.strings="",row.names=NULL,quote="",comment.char="", stringsAsFactors = FALSE, check.names = FALSE)
rownames(hm450.anno) <- hm450.anno$probeID
hm450.anno.unmasked <- hm450.anno[hm450.anno$MASK.general=='FALSE',]
rm(hm450.anno)
hm450.anno.epig <-read.table("hm450.manifest.EpigeneticallySilence_Sort.tsv",header=T,sep="\t",na.strings="NA",row.names=NULL,quote="",comment.char="", stringsAsFactors = FALSE, check.names = FALSE)
## hm450.manifest.EpigeneticallySilence_Sort.tsv is the file for CpG sites with distance from TSS. I generated this file from GDC harmonized DNA methylation file for genCode22
hm450.anno.epig <- hm450.anno.epig[which(hm450.anno.epig$Position_to_TSS!='.'),]
hm450.anno.epig.tss1500 <- hm450.anno.epig[abs(as.numeric(hm450.anno.epig$Position_to_TSS)) <= 1500,]
hm450.anno.epig.tss1500.unmasked <- hm450.anno.epig.tss1500[hm450.anno.epig.tss1500$ProbeID%in%hm450.anno.unmasked$probeID,]
hm450.tss1500 <- hm450.anno.epig.tss1500.unmasked[,c("ProbeID", "Transcript_ID", "Gene_Symbol","Position_to_TSS", "Chromosome")]
colnames(hm450.tss1500) <- gsub("Chromosome","Chr", colnames(hm450.tss1500))
#TCGA.Harmonized <- read.table("TCGA_GDC_Harmonided_RNASeq-GFT.txt",header=T,sep="\t",na.strings="NA",row.names=NULL,quote="",comment.char="", stringsAsFactors = FALSE, check.names = FALSE)
TCGA.Harmonized <- read.table("TCGA_GDC_Harmonided_Uniq_GTF.txt",header=F,sep="\t",na.strings="NA",row.names=NULL,quote="",comment.char="", stringsAsFactors = FALSE, check.names = FALSE)
colnames(TCGA.Harmonized) <- c("Chr", "Reference", "Transcript", "Start", "End", "Strand", "Transcript_ID", "Gene_Symbol")
masked <- which(TCGA.Harmonized$Chr%in% c("chrX", "chrY", "chrM"))
data <- read.table("TCGA_GDC_Harmonided_RNASeq-GFT.txt",header=T,sep="\t",na.strings="",row.names=NULL,quote="",comment.char="")
TCGA.Har.Masked <- TCGA.Harmonized[-masked,]
TCGA.Har.Masked <- data[data$ENSB_ID %in% TCGA.Har.Masked$Transcript_ID,]
colnames(TCGA.Har.Masked) <- gsub("ENSB_ID", "Transcript_ID", colnames(TCGA.Har.Masked))
TCGA.Har.Masked <- TCGA.Har.Masked[,c("Chr", "Transcript_ID", "Gene_Symbol", "Gene_status")]
masked <- which(hm450.tss1500$Chr%in% c("chrX", "chrY", "chrM"))
hm450.tss1500.Masked <- hm450.tss1500[-masked,]
hm450.tss1500.Masked.Uniq <- unique(hm450.tss1500.Masked[,c("ProbeID", "Gene_Symbol", "Chr", "Transcript_ID")]) ## Unique gene and probe within +/- 1500 from TSS

######## DNA methylation and expression data ##########
Meth <- CHOL.Meth; colnames(Meth) <- substr(colnames(Meth), 1, 16)
Exp <- CHOL.FPKM; colnames(Exp) <- substr(colnames(Exp), 1, 16)
TumorID <- intersect(colnames(Meth[grep("-01A", colnames(Meth))]), colnames(Exp[grep("-01A", colnames(Exp))]))
NormalID <- intersect(colnames(Meth[grep("-11A", colnames(Meth))]), colnames(Exp[grep("-11A", colnames(Exp))]))
Meth <- Meth[,c(NormalID, TumorID)]; Exp <- Exp[,c(NormalID, TumorID)]
Meth$ProbeID <- rownames(Meth)
Meth <- na.omit(Meth) # #Meth <- Meth[complete.cases(Meth),] # Like above command this will also remove all "NA" masked data
Meth.tss1500 <- merge(hm450.tss1500.Masked.Uniq, Meth, by = "ProbeID") 
#Meth.tss1500$N.Mean <- rowMeans(Meth.tss1500[,grep("-11A", colnames(Meth.tss1500))])#Mean Beta Normal
#Meth.tss1500$T.Mean <- rowMeans(Meth.tss1500[,grep("-01A", colnames(Meth.tss1500))])#Mean Beta Tumor
#Meth.tss1500$Delta.Beta <- Meth.tss1500$T.Mean - Meth.tss1500$N.Mean
#Meth.tss1500$FoldChange <- Meth.tss1500$T.Mean/Meth.tss1500$N.Mean
Exp <- log2(Exp+1); Exp$Transcript_ID <- rownames(Exp)
Exp.Masked <- merge(Exp, TCGA.Har.Masked, by = "Transcript_ID")
Exp.Masked.T <- Exp.Masked[,grep("-01A", colnames(Exp.Masked))]
Exp.Masked.N <- Exp.Masked[,grep("-11A", colnames(Exp.Masked))]
Exp.Masked.New <- cbind(Exp.Masked.N, cbind(Exp.Masked.T, Exp.Masked$Gene_Symbol))
colnames(Exp.Masked.New) <- gsub("Exp.Masked\\$", "", colnames(Exp.Masked.New))
Exp.Masked.New$Gene_Symbol <- as.character(Exp.Masked.New$Gene_Symbol)
#Exp.Masked$N.Mean <- rowMeans(Exp.Masked.N) ## Mean gene expression in normal
#Exp.Masked$T.Mean <- rowMeans(Exp.Masked.T) ## Mean gene expression in tumor
#Exp.Masked$FoldChange <- Exp.Masked$T.Mean/Exp.Masked$N.Mean
Merged_data <- merge(Meth.tss1500, Exp.Masked.New, by="Gene_Symbol")
Merged_data.Meth <- Merged_data[,grep("A.x", colnames(Merged_data))]
Merged_data.Exp <- Merged_data[,grep("A.y", colnames(Merged_data))]

keep <- which(rowSums(Merged_data.Exp == 0) >= 36)

#keep <- rowSums(Merged_data.Exp == 0) >= 36, ])
Merged_data.Meth <- Merged_data.Meth[-keep,]
Merged_data.Exp <- Merged_data.Exp[-keep,]
Merged_data.New <- cbind(Merged_data.Meth, Merged_data.Exp)
################## student t-test #####################
#Meth.tss1500.T <- Meth.tss1500[,grep("-01A", colnames(Meth.tss1500))]
#Meth.tss1500.N <- Meth.tss1500[,grep("-11A", colnames(Meth.tss1500))]
ll <- mapply(function(x,y)cor.test(as.numeric(Merged_data.Meth[x,]),as.numeric(Merged_data.Exp[y,]), method = "spearman"),
             1:nrow(Merged_data.Exp),
             1:nrow(Merged_data.Exp),
             SIMPLIFY=FALSE) ## Few have only one or two value, Pearson's correlation create problem for them
Merged_data.New$p.value <- sapply(ll,'[[','p.value')
rho <- sapply(ll,'[[','estimate')
Merged_data.New$rho <- unname(rho)
Merged_data.New$adj.p.value <- p.adjust(Merged_data.New$p.value, "BH")
anno <- Merged_data[-keep, 1:4]
Merged_data.New <- cbind(anno, Merged_data.New)

Merged_data.New1 <- Merged_data.New[, c("Gene_Symbol", "ProbeID", "p.value", "rho", "adj.p.value")]
Min.rho <- aggregate(rho ~ Gene_Symbol, Merged_data.New1, min)
Min.rho.min <- unique(merge(Min.rho, Merged_data.New1))

Max.rho <- aggregate(rho ~ Gene_Symbol, Merged_data.New1, max)
Max.rho.max <- unique(merge(Max.rho, Merged_data.New1))
tmp <- merge(Min.rho.min, Max.rho.max, by="Gene_Symbol")
tmp <- tmp[!duplicated(tmp$Gene_Symbol),]
tmp1 <- tmp[,c("rho.x", "rho.y")]
colnames(tmp1) <- c("rho_neg", "rho_pos")
tmp$max <- colnames(tmp1)[apply(abs(tmp1),1,which.max)]

tmp.1 <- tmp[grep("rho_neg", tmp$max),c(grep("Gene_Symbol|\\.x|max", colnames(tmp)))]
tmp.2 <- tmp[grep("rho_pos", tmp$max),c(grep("Gene_Symbol|\\.y|max", colnames(tmp)))]
colnames(tmp.1) <- gsub("\\.x|\\.y","", colnames(tmp.1))
colnames(tmp.2) <- gsub("\\.x|\\.y","", colnames(tmp.2))
tmp2 <- rbind(tmp.1, tmp.2)
tmp2 <- merge(tmp2, TCGA.Har.Masked, by="Gene_Symbol")
tmp2 <- tmp2[,c("Gene_Symbol", "ProbeID", "rho", "p.value", "adj.p.value",     "max",   "Chr",      "Transcript_ID",    "Gene_status")]
############################################################
library(dplyr)
tmp3 <- na.omit(Merged_data.New %>% left_join(tmp2, by=c("Gene_Symbol", "ProbeID")))
tmp3 <- tmp3[!duplicated(tmp3),]

dat.Me <- tmp3[, grep("Gene|ProbeID|p.value|Transcript_ID|rho|A\\.x", colnames(tmp3))]
colnames(dat.Me) <- gsub(".x", "", colnames(dat.Me))
dat.Me <- dat.Me[!duplicated(dat.Me[,c('Gene_Symbol','ProbeID')]),] ## Make unique data.frame based on Gene_Symbol and ProbeID only
dat.Me.T <- dat.Me[,grep("-01A", colnames(dat.Me))]
dat.Me.N <- dat.Me[,grep("-11A", colnames(dat.Me))]

Meth.T30 <- vector("numeric", length = nrow(dat.Me))
Meth.T70 <- vector("numeric", length = nrow(dat.Me))
Meth.N30 <- vector("numeric", length = nrow(dat.Me))
Meth.N70 <- vector("numeric", length = nrow(dat.Me))

## This part is slow
# for(i in 1:nrow(dat.Me.T)){
#   tmp <- round(quantile(dat.Me.T[i,], probs=c(0.30,0.7), na.rm=TRUE, names=FALSE),3)
#   Meth.T30[i] <- tmp[1]
#   Meth.T70[i] <- tmp[2]
#   tmp <- round(quantile(dat.Me.N[i,], probs=c(0.30,0.7), na.rm=TRUE, names=FALSE),3)
#   Meth.N30[i] <- tmp[1]
#   Meth.N70[i] <- tmp[2]
# }

## This is the same code as above in for loop, but it's very fast compare to for loop
ll <- apply(dat.Me.N, 1, quantile, probs = c(0.30, 0.70),  na.rm = TRUE, names=FALSE)
N30 <- round(ll[1,],3)
N70 <- round(ll[2,], 3)

ll <- apply(dat.Me.T, 1, quantile, probs = c(0.30, 0.70),  na.rm = TRUE, names=FALSE)
T30 <- round(ll[1,],3)
T70 <- round(ll[2,], 3)
dat.Me <- cbind(dat.Me, N30, N70, T30, T70)

