## R code for Epigenetic silencing analysis is based on the methodology proposed by Noushmehr et. al. in TCGA hepatocellular carcinoma, Cell (2017), 169, 1237-1241
## I requested multiple times but Hui Shen didn't prove it. So I make this code for Epigenetic silencing analysis
## For this analysis we need gene expression and DNA methylation data for both tumor and normal.
## I can make changes if there are no normal samples. But in the best case scenario, we need both case and control data.
setwd("F:/OneDrive - University of Nebraska Medical Center/Epigenetic Silencing/")
library(pacman)
p_load(multtest, matrixStats, plyr)
#######################################################
load("CHOL.RData")
rm(list = ls()[!ls() %in% c("CHOL.Meth", "CHOL.FPKM", "CHOL.miRNA.RPM", "CHOL.FPKM.UQ", "CHOL.HTSeq.Counts")])
## hm450.manifest.hg38.tsv is the annotation file generated by Hui Shen, NAR, 2016 (Comprehensive chracterization, annotation and innovative use of infinium DNA methylation BeadChip)  
hm450.anno <-read.table("hm450.manifest.hg38.tsv",header=T,sep="\t",na.strings="",row.names=NULL,quote="",comment.char="", stringsAsFactors = FALSE, check.names = FALSE)
rownames(hm450.anno) <- hm450.anno$probeID
hm450.anno.unmasked <- hm450.anno[hm450.anno$MASK.general=='FALSE',]
rm(hm450.anno)
hm450.anno.epig <-read.table("hm450.manifest.EpigeneticallySilence_Sort.tsv",header=T,sep="\t",na.strings="NA",row.names=NULL,quote="",comment.char="", stringsAsFactors = FALSE, check.names = FALSE)
## hm450.manifest.EpigeneticallySilence_Sort.tsv is the file for CpG sites with distance from TSS. I generated this file from GDC harmonized DNA methylation file for genCode22
hm450.anno.epig <- hm450.anno.epig[which(hm450.anno.epig$Position_to_TSS!='.'),]
hm450.anno.epig.tss1500 <- hm450.anno.epig[abs(as.numeric(hm450.anno.epig$Position_to_TSS)) <= 1500,]
hm450.anno.epig.tss1500.unmasked <- hm450.anno.epig.tss1500[hm450.anno.epig.tss1500$ProbeID%in%hm450.anno.unmasked$probeID,]
hm450.tss1500 <- hm450.anno.epig.tss1500.unmasked[,c("ProbeID", "Transcript_ID", "Gene_Symbol","Position_to_TSS", "Chromosome")]
colnames(hm450.tss1500) <- gsub("Chromosome","Chr", colnames(hm450.tss1500))
#TCGA.Harmonized <- read.table("TCGA_GDC_Harmonided_RNASeq-GFT.txt",header=T,sep="\t",na.strings="NA",row.names=NULL,quote="",comment.char="", stringsAsFactors = FALSE, check.names = FALSE)
TCGA.Harmonized <- read.table("TCGA_GDC_Harmonided_Uniq_GTF.txt",header=F,sep="\t",na.strings="NA",row.names=NULL,quote="",comment.char="", stringsAsFactors = FALSE, check.names = FALSE)
colnames(TCGA.Harmonized) <- c("Chr", "Reference", "Transcript", "Start", "End", "Strand", "Transcript_ID", "Gene_Symbol")
masked <- which(TCGA.Harmonized$Chr%in% c("chrX", "chrY", "chrM"))
TCGA.Har.Masked <- TCGA.Harmonized[-masked,]
TCGA.Har.Masked <- TCGA.Har.Masked[,c("Chr", "Transcript_ID", "Gene_Symbol")]
masked <- which(hm450.tss1500$Chr%in% c("chrX", "chrY", "chrM"))
hm450.tss1500.Masked <- hm450.tss1500[-masked,]
hm450.tss1500.Masked.Uniq <- unique(hm450.tss1500.Masked[,c("ProbeID", "Gene_Symbol", "Chr")]) ## Unique gene and probe within +/- 1500 from TSS

######## DNA methylation and expression data ##########
Meth <- CHOL.Meth; colnames(Meth) <- substr(colnames(Meth), 1, 16)
Exp <- CHOL.FPKM; colnames(Exp) <- substr(colnames(Exp), 1, 16)
TumorID <- intersect(colnames(Meth[grep("-01A", colnames(Meth))]), colnames(Exp[grep("-01A", colnames(Exp))]))
NormalID <- intersect(colnames(Meth[grep("-11A", colnames(Meth))]), colnames(Exp[grep("-11A", colnames(Exp))]))
Meth <- Meth[,c(NormalID, TumorID)]; Exp <- Exp[,c(NormalID, TumorID)]
Meth$ProbeID <- rownames(Meth)
Meth <- na.omit(Meth) # #Meth <- Meth[complete.cases(Meth),] # Like above command this will also remove all "NA" masked data
Meth.tss1500 <- merge(hm450.tss1500.Masked.Uniq, Meth, by = "ProbeID") 
Meth.tss1500$N.Mean <- rowMeans(Meth.tss1500[,grep("-11A", colnames(Meth.tss1500))])#Mean Beta Normal
Meth.tss1500$T.Mean <- rowMeans(Meth.tss1500[,grep("-01A", colnames(Meth.tss1500))])#Mean Beta Tumor
Meth.tss1500$Delta.Beta <- Meth.tss1500$T.Mean - Meth.tss1500$N.Mean
Meth.tss1500$FoldChange <- Meth.tss1500$T.Mean/Meth.tss1500$N.Mean

################## student t-test #####################
Meth.tss1500.T <- Meth.tss1500[,grep("-01A", colnames(Meth.tss1500))]
Meth.tss1500.N <- Meth.tss1500[,grep("-11A", colnames(Meth.tss1500))]
ll <- mapply(function(x,y)t.test(Meth.tss1500.T[x,],Meth.tss1500.N[y,]),
             1:nrow(Meth.tss1500.T),
             1:nrow(Meth.tss1500.N),
             SIMPLIFY=FALSE) ## Few have only one or two value, Pearson's correlation create problem for them
Meth.tss1500$p.value.Meth <- sapply(ll,'[[','p.value')
Meth.tss1500$adj.p.value.Meth <- p.adjust(Meth.tss1500$p.value.Meth, "BH")

Exp <- log2(Exp+1); Exp$Transcript_ID <- rownames(Exp)
Exp.Masked <- merge(Exp, TCGA.Har.Masked, by = "Transcript_ID")
Exp.Masked.T <- Exp.Masked[,grep("-01A", colnames(Exp.Masked))]
Exp.Masked.N <- Exp.Masked[,grep("-11A", colnames(Exp.Masked))]
Exp.Masked$N.Mean <- rowMeans(Exp.Masked.N) ## Mean gene expression in normal
Exp.Masked$T.Mean <- rowMeans(Exp.Masked.T) ## Mean gene expression in tumor
Exp.Masked$FoldChange <- Exp.Masked$T.Mean/Exp.Masked$N.Mean

ll <- mapply(function(x,y)t.test(Exp.Masked.T[x,],Exp.Masked.N[y,]),
             1:nrow(Exp.Masked.T),
             1:nrow(Exp.Masked.N),
             SIMPLIFY=FALSE) ## Few have only one or two value, Pearson's correlation create problem for them
Exp.Masked$p.value.Meth <- sapply(ll,'[[','p.value')
Exp.Masked$adj.p.value.Meth <- p.adjust(Exp.Masked$p.value.Meth, "BH")

######################################################
Meth.tss1500.Delta <- Meth.tss1500[Meth.tss1500$adj.p.value.Meth < 0.0001 & Meth.tss1500$Delta.Beta > 0.1,]
Exp.Masked.DownReg <- Exp.Masked[Exp.Masked$adj.p.value.Meth < 0.0001 & Exp.Masked$FoldChange < 0.5,]

dat.meth.exp <- merge(Meth.tss1500.Delta, Exp.Masked.DownReg, by = "Gene_Symbol")
######################################################
dat.Me <- dat.meth.exp[,grep("A.x", colnames(dat.meth.exp))]
dat.Ex <- dat.meth.exp[,grep("A.y", colnames(dat.meth.exp))]
colnames(dat.Me) <- gsub(".x", "", colnames(dat.Me))
colnames(dat.Ex) <- gsub(".y", "", colnames(dat.Ex))
dat.Me.T <- dat.Me[,grep("-01A", colnames(dat.Me))]
dat.Me.N <- dat.Me[,grep("-11A", colnames(dat.Me))]
dat.Ex.T <- dat.Ex[,grep("-01A", colnames(dat.Ex))]
dat.Ex.N <- dat.Ex[,grep("-11A", colnames(dat.Ex))]

T.Hyper <- vector("numeric", length = dim(dat.Ex)[1])
T.Hypo <- vector("numeric", length = dim(dat.Ex)[1])
Hyper.Exp.Mean <- vector("numeric", length = dim(dat.Ex)[1])
Hypo.Exp.Mean <- vector("numeric", length = dim(dat.Ex)[1])

for(i in 1:nrow(dat.Me.T)){
  Hyper <- which(dat.Me.T[i,] >= 0.3)
  Hypo <- which(dat.Me.T[i,] < 0.3)
  T.Hyper[i] <- length(Hyper)
  T.Hypo[i] <- length(Hypo)
  Hyper.Exp.Mean[i] <- mean(as.numeric(dat.Ex.T[i, Hyper]))
  Hypo.Exp.Mean[i] <- mean(as.numeric(dat.Ex.T[i, Hypo]), na.rm = TRUE)
}

dat.meth.exp$T.Hyper <- T.Hyper
dat.meth.exp$T.Hypo <- T.Hypo
dat.meth.exp$Hyper.Exp.Mean <- Hyper.Exp.Mean
dat.meth.exp$Hypo.Exp.Mean <- Hypo.Exp.Mean
dat.meth.exp$Mean.Diff <-  Hypo.Exp.Mean - Hyper.Exp.Mean
dat.meth.exp$freq <- T.Hyper/(T.Hyper+T.Hypo)
dat.meth.exp <- na.omit(dat.meth.exp)

########### Select Epigenetic silenced genes #########
Epigenetic_silenced_Noushmer <- ddply(rdat.meth.exp, .(Gene_Symbol), function(x) x[which.max(x$Mean.Diff),])
Epigenetic_silenced_Noushmer <- Epigenetic_silenced_Noushmer[Epigenetic_silenced_Noushmer$Mean.Diff > 0, ] ## Gene which are only downregulated
rownames(Epigenetic_silenced_Noushmer) <- Epigenetic_silenced_Noushmer$Gene_Symbol

rm(list = ls()[!ls() %in% c("Epigenetic_silenced_Noushmer", "dat.meth.exp","CHOL.Meth", "CHOL.FPKM", "CHOL.FPKM.UQ")])
save.image(file = "EpigeneticSilence_Noushmer_Nov-02.2017.RData")
